### PR TITLE
FIXES #2640: LucenePartitioner.getNextNewerPartitionInfo() sometimes …

### LINKED ISSUE #2640: LucenePartitioner.getNextNewerPartitionInfo() sometimes returns invalid (null) result

### PR DESCRIPTION
**Pull request by wshaib-apple**
…returns invalid (null) result

### CODE REVIEW

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 1]
```diff
@@ -1337,6 +1341,170 @@ class GroupSpec {
...
+                        case GREATER_THAN_OR_EQUALS:
+                            return value >= comparand;
+                        default:
+                            return false;
+                    }
+                })
+                .collect(Collectors.toList());
+
+        if (sortType == SortType.ASCENDING) {
+            hits.sort((Map.Entry<Tuple, Long> a, Map.Entry<Tuple, Long> b) -> {
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1489]: Not a big deal, but I think these sorts could be simplified using something like:
```
Comparator.comparing(entry -> entry.getValue()).thenComparing(Comparator.comparing(entry -> entry.getKey()))
```

And you can use `comparator.reversed()` to do descending.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 2]
```diff
@@ -1337,6 +1341,170 @@ class GroupSpec {
...
+                int c = a.getValue().compareTo(b.getValue());
+                if (c == 0) {
+                    return a.getKey().compareTo(b.getKey());
+                }
+                return c;
+            });
+        } else {
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1496]: You could limit this sorting to `DESCENDING`, and skip the sorting in the case of `UNSORTED`

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java (hunk 2 above)
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1496]: `UNSORTED` behaves like `DESCENDING` (referring to the partition scanning direction)

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java [hunk 3]
```diff
@@ -1337,6 +1341,158 @@ class GroupSpec {
...
+        final Sort sort;
+        if (sortType == SortType.UNSORTED) {
+            sort = null;
+        } else {
+            sort = new Sort(new SortField(isSynthetic ? "complex_timestamp" : "timestamp", SortField.Type.LONG, sortType == SortType.DESCENDING));
+        }
+
+        final List<LuceneQueryClause> luceneQueryClauses;
+
+        if (comparisonType == Type.NOT_EQUALS) {
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1359]: This is not actually used in the test.

**Review by ScottDugas — APPROVED** [review]

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java (hunk 3 above)
Comment (wshaib-apple) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:1359]: it is used by the other unit test that calls `buildLuceneQuery` in the other PR, from which this method was extracted.

**Pull request merged**